runtime/ms-rest-azure/lib/login.js

```diff
@@ -468,4 +470,232 @@ exports.withServicePrincipalSecretWithAuthResponse = function withServicePrincip
+  }
+  if (!credsObj.sqlManagementEndpointUrl) {
+    throw new Error(`"sqlManagementEndpointUrl" is missing from the auth file: ${filePath}.`);
+  }
+  if (!credsObj.galleryEndpointUrl) {
+    throw new Error(`"galleryEndpointUrl" is missing from the auth file: ${filePath}.`);
+  }
+  if (!credsObj.managementEndpointUrl) {
+    throw new Error(`"managementEndpointUrl" is missing from the auth file: ${filePath}.`);
+  }
+}
+
+function _foundManagementEndpointUrl(authFileUrl, envUrl) {
+  if (!authFileUrl || (authFileUrl && typeof authFileUrl.valueOf() !== 'string')) {
+    throw new Error('authFileUrl cannot be null or undefined and must be of type string.');
+  }
+
+  if (!envUrl || (envUrl && typeof envUrl.valueOf() !== 'string')) {
+    throw new Error('envUrl cannot be null or undefined and must be of type string.');
+  }
+
+  authFileUrl = authFileUrl.endsWith('/') ? authFileUrl.slice(0, -1) : authFileUrl;  
+  envUrl = envUrl.endsWith('/') ? envUrl.slice(0, -1) : envUrl;
+  return (authFileUrl.toLowerCase() === envUrl.toLowerCase());
+}
+
+/**
+ * Authenticates using the service principal information provided in the auth file. This method will set 
+ * the subscriptionId from the auth file to the user provided environment variable in the options 
+ * parameter or the default AZURE_SUBSCRIPTION_ID.
+ * @param {object} [options] - Optional parameters
+ * @param {string} [options.filePath] - Absolute file path to the auth file. If not provided 
+ * then please set the environment variable AZURE_AUTH_LOCATION.
+ * @param {string} [options.subscriptionEnvVariableName] - The subscriptionId environment variable 
+ * name. Default is 'AZURE_SUBSCRIPTION_ID'.
+ * @param {function} callback - The callback
+ */
+function _withAuthFile(options, callback) {
+  if (!callback && typeof options === 'function') {
+    callback = options;
+    options = {};
+  }
+  if (!options) options = { filePath: '' };
+  if (!callback) {
+    throw new Error('callback cannot be null or undefined.');
+  }
+
+  let filePath = options.filePath;
+  let authLocation = process.env[azureConstants.AZURE_AUTH_LOCATION];
+  let subscriptionEnvVariableName = options.subscriptionEnvVariableName || 'AZURE_SUBSCRIPTION_ID';
+  if (!filePath) {
+    if (!authLocation) {
+      let msg = `Either provide an absolute file path to the auth file or set/export the environment variable - ${azureConstants.AZURE_AUTH_LOCATION}.`;
+      return callback(new Error(msg));
+    }
+    filePath = authLocation;
+  }
+  //expand ~ to user's home directory.
+  if(filePath.startsWith('~')) {
+    filePath = msRest.homeDir(filePath.slice(1));
+  }
+
+  let content = null, credsObj = {}, optionsForSpSecret = {};
+  try {
+    content = fs.readFileSync(filePath, {encoding: 'utf8'});
+    credsObj = JSON.parse(content);
+    _validateAuthFileContent(credsObj, filePath);
+    if (!process.env[subscriptionEnvVariableName]) {
```

Comment:
The environment variable that we are setting over here is the env. variable for subscriptionId. It is not the env. var for auth location. Node sdk has a recommended set of environment variables for things like subscriptionId, clientId, username, password, domain, etc.
So in this case, if the user provides his custom environment variable name that is being used in his app then we use that env. variable to set the value of subscriptionId. IF not then we set the subscriptionId to AZURE_SUBSCRIPTION_ID env. variable.
This subscription Id will be the only subscriptionId that the customer can use while instantiating the client since the SP is linked to that subscription. Using any other subscriptionId will be setting him up for failure.

I know this is a bit confusing. But whatever I said above, does that make sense or clear the initial question that you asked?